In [ ]:
from sklearn.model_selection import train_test_split
from sklearn.metrics import mean_squared_error
from sklearn.preprocessing import LabelBinarizer

import pandas as pd
import numpy as np

In [ ]:
# Baixar os dados de treino
!wget https://www.dropbox.com/s/exhzlbwyipz9tio/train.csv

treino = pd.read_csv("train.csv", index_col=0)

--2021-10-28 05:03:08--  https://www.dropbox.com/s/exhzlbwyipz9tio/train.csv
Resolving www.dropbox.com (www.dropbox.com)... 162.125.6.18, 2620:100:601d:18::a27d:512
Connecting to www.dropbox.com (www.dropbox.com)|162.125.6.18|:443... connected.
HTTP request sent, awaiting response... 301 Moved Permanently
Location: /s/raw/exhzlbwyipz9tio/train.csv [following]
--2021-10-28 05:03:08--  https://www.dropbox.com/s/raw/exhzlbwyipz9tio/train.csv
Reusing existing connection to www.dropbox.com:443.
HTTP request sent, awaiting response... 302 Found
Location: https://uc8ecc880578ef6791625ca54d23.dl.dropboxusercontent.com/cd/0/inline/BY05LfBhRuC1kKR6HMdjMZaz9t0bpVwJlL7c-ZAXzcccfwwk8Wfhhy54TSBwfFNfMo6ZKCnmLhOW3wiiFxv-86t5T3QWw1JnGHFrDpV-kOUipPMkItZ1-B8EaKwLs8Hodfqc9HxHu78UjHeKKkjQoVqx/file# [following]
--2021-10-28 05:03:08--  https://uc8ecc880578ef6791625ca54d23.dl.dropboxusercontent.com/cd/0/inline/BY05LfBhRuC1kKR6HMdjMZaz9t0bpVwJlL7c-ZAXzcccfwwk8Wfhhy54TSBwfFNfMo6ZKCnmLhOW3wiiFxv-86t5T3QWw1JnGHF

In [ ]:
# Baixar os dados de teste
!wget https://www.dropbox.com/s/8k1ob4bumn9sxvn/test.csv

teste = pd.read_csv("test.csv", index_col=0)

--2021-10-28 05:04:13--  https://www.dropbox.com/s/8k1ob4bumn9sxvn/test.csv
Resolving www.dropbox.com (www.dropbox.com)... 162.125.4.18, 2620:100:6030:18::a27d:5012
Connecting to www.dropbox.com (www.dropbox.com)|162.125.4.18|:443... connected.
HTTP request sent, awaiting response... 301 Moved Permanently
Location: /s/raw/8k1ob4bumn9sxvn/test.csv [following]
--2021-10-28 05:04:13--  https://www.dropbox.com/s/raw/8k1ob4bumn9sxvn/test.csv
Reusing existing connection to www.dropbox.com:443.
HTTP request sent, awaiting response... 302 Found
Location: https://uc402daf11f31781b4a20e680dd3.dl.dropboxusercontent.com/cd/0/inline/BY1ORC84thphieJt6MOrHous0MBnSp-gTO3OSpcW3GSY_HX5wPATNSsivAULQL0y4I56NKx-1_05KH_cKU7UcJaFZYgNvJeu_AlOQrKXFea9FHYhBtR7t--oVCIdBPj4MTXomj8XMyFCCEoZ69H_tfsq/file# [following]
--2021-10-28 05:04:13--  https://uc402daf11f31781b4a20e680dd3.dl.dropboxusercontent.com/cd/0/inline/BY1ORC84thphieJt6MOrHous0MBnSp-gTO3OSpcW3GSY_HX5wPATNSsivAULQL0y4I56NKx-1_05KH_cKU7UcJaFZYgNvJeu_AlOQ

In [ ]:
print("Shape of train: ", treino.shape)
print("Shape of test: ", teste.shape)

Shape of train:  (1460, 80)
Shape of test:  (1459, 79)


In [ ]:
treino.head(5)

MSSubClass MSZoning  LotFrontage  ...  SaleType SaleCondition SalePrice
Id                                    ...                                  
1           60       RL         65.0  ...        WD        Normal    208500
2           20       RL         80.0  ...        WD        Normal    181500
3           60       RL         68.0  ...        WD        Normal    223500
4           70       RL         60.0  ...        WD       Abnorml    140000
5           60       RL         84.0  ...        WD        Normal    250000

[5 rows x 80 columns]

In [ ]:
teste.head(5)

MSSubClass MSZoning  LotFrontage  ...  YrSold SaleType SaleCondition
Id                                      ...                               
1461          20       RH         80.0  ...    2010       WD        Normal
1462          20       RL         81.0  ...    2010       WD        Normal
1463          60       RL         74.0  ...    2010       WD        Normal
1464          60       RL         78.0  ...    2010       WD        Normal
1465         120       RL         43.0  ...    2010       WD        Normal

[5 rows x 79 columns]

## Preparação de dados:
### Elementos faltantes
Em primeiro lugar, precisamos saber se existem e onde estão os elementos faltantes ( Not a Number [NaN] ). Se eles devem ser preenchidos ou retirados, e quantos são.

In [ ]:
pd.set_option('display.max_rows', None) ## mostrar todas as linhas
treino.isnull().sum()

MSSubClass          0
MSZoning            0
LotFrontage       259
LotArea             0
Street              0
Alley            1369
LotShape            0
LandContour         0
Utilities           0
LotConfig           0
LandSlope           0
Neighborhood        0
Condition1          0
Condition2          0
BldgType            0
HouseStyle          0
OverallQual         0
OverallCond         0
YearBuilt           0
YearRemodAdd        0
RoofStyle           0
RoofMatl            0
Exterior1st         0
Exterior2nd         0
MasVnrType          8
MasVnrArea          8
ExterQual           0
ExterCond           0
Foundation          0
BsmtQual           37
BsmtCond           37
BsmtExposure       38
BsmtFinType1       37
BsmtFinSF1          0
BsmtFinType2       38
BsmtFinSF2          0
BsmtUnfSF           0
TotalBsmtSF         0
Heating             0
HeatingQC           0
CentralAir          0
Electrical          1
1stFlrSF            0
2ndFlrSF            0
LowQualFinSF        0
GrLivArea 

Assim, podemos ver que as colunas:

   ▪ Alley  tem  1369 dados faltantes.
 
   ▪ PoolQC tem        1453 dados faltantes.

   ▪ Fence tem 1179 dados faltantes.

   ▪ MiscFeature  tem  1406 dados faltantes.
 
 A proporção de valores NaNs nessas colunas é exorbitante. É mais coeso retirá-las.

In [ ]:
pd.reset_option('display.max_columns')
pd.reset_option('display.max_rows') ## parar de mostrar todas as linhas

treino = treino.drop('Alley',axis=1)
treino = treino.drop('PoolQC',axis=1)
treino = treino.drop('Fence',axis=1)
treino = treino.drop('MiscFeature',axis=1)
treino.columns

Index(['MSSubClass', 'MSZoning', 'LotFrontage', 'LotArea', 'Street',
       'LotShape', 'LandContour', 'Utilities', 'LotConfig', 'LandSlope',
       'Neighborhood', 'Condition1', 'Condition2', 'BldgType', 'HouseStyle',
       'OverallQual', 'OverallCond', 'YearBuilt', 'YearRemodAdd', 'RoofStyle',
       'RoofMatl', 'Exterior1st', 'Exterior2nd', 'MasVnrType', 'MasVnrArea',
       'ExterQual', 'ExterCond', 'Foundation', 'BsmtQual', 'BsmtCond',
       'BsmtExposure', 'BsmtFinType1', 'BsmtFinSF1', 'BsmtFinType2',
       'BsmtFinSF2', 'BsmtUnfSF', 'TotalBsmtSF', 'Heating', 'HeatingQC',
       'CentralAir', 'Electrical', '1stFlrSF', '2ndFlrSF', 'LowQualFinSF',
       'GrLivArea', 'BsmtFullBath', 'BsmtHalfBath', 'FullBath', 'HalfBath',
       'BedroomAbvGr', 'KitchenAbvGr', 'KitchenQual', 'TotRmsAbvGrd',
       'Functional', 'Fireplaces', 'FireplaceQu', 'GarageType', 'GarageYrBlt',
       'GarageFinish', 'GarageCars', 'GarageArea', 'GarageQual', 'GarageCond',
       'PavedDrive', 'WoodDeckSF'

In [ ]:
from numpy import isnan
# Em quais colunas estão dados null?
print(treino.columns[treino.isnull().any()])

Index(['LotFrontage', 'MasVnrType', 'MasVnrArea', 'BsmtQual', 'BsmtCond',
       'BsmtExposure', 'BsmtFinType1', 'BsmtFinType2', 'Electrical',
       'FireplaceQu', 'GarageType', 'GarageYrBlt', 'GarageFinish',
       'GarageQual', 'GarageCond'],
      dtype='object')


### NaNs da coluna "LotFrontage":

É uma coluna numérica. Assim, podemos preencher esse valores de diversas formas, como utilizar o comando .fillna(); com média, moda, mediana dos valores que possuímos;  ou o KNN. Neste dataset, vamos usar o KNN.

In [ ]:
pd.reset_option('display.max_rows')
print("Temos NaNs? {}\n".format(treino['LotFrontage'].isnull().any()))
# mostrando todos os NaNs
df = pd.DataFrame(treino,columns=['LotFrontage'])
print(df[df['LotFrontage'].isna()])

Temos NaNs? True

      LotFrontage
Id               
8             NaN
13            NaN
15            NaN
17            NaN
25            NaN
...           ...
1430          NaN
1432          NaN
1442          NaN
1444          NaN
1447          NaN

[259 rows x 1 columns]


#### Utilizando KNN:
O scikit-learn tem um imputador de valores de NaN que utiliza da média dos N vizinhos próximos, [encontrado neste link.](https://scikit-learn.org/stable/modules/generated/sklearn.impute.KNNImputer.html)

Ele tem alguns parametros básicos:

•*missing_value*:O tipo do valor que você quer  substituir.

•*n_neighborsint*: A quantidade de vizinhos para calcular a média, o padrão é 5.

•*weights{‘uniform’, ‘distance’}or callable*: O peso de cada vizinho para a média, o padrão é todos com o mesmo peso. 

•*metric{‘nan_euclidean’} or callable*: Métrica das distâncias entre os vizinhos, o padrão é nan_euclidean.

In [ ]:
from sklearn.impute import KNNImputer

#knn:
imp = KNNImputer(n_neighbors=5, weights='uniform', metric='nan_euclidean')

'''Assim, podemos transformar o dataset. Entretanto o KNN é normalmente utilizado em 
dimensão 2, e como uma coluna tem dimensão 1, usamos o .ravel() para o achatamento'''
treino["LotFrontage"] = imp.fit_transform(treino[["LotFrontage"]]).ravel()

print("Temos NaNs na 'LotFrontage'? {}\n".format(treino['LotFrontage'].isnull().any()))

#coluna sem NaNs:
print(treino['LotFrontage'])

Temos NaNs na 'LotFrontage'? False

Id
1       65.0
2       80.0
3       68.0
4       60.0
5       84.0
        ... 
1456    62.0
1457    85.0
1458    66.0
1459    68.0
1460    75.0
Name: LotFrontage, Length: 1460, dtype: float64


### NaNs da coluna "FireplaceQu":

A coluna 'FireplaceQu', diferentemente da 'LotFrontage', é composta por strings. Apesar da moda impactar na distribuição dos dados, não temos como aplicar média ou mediana neste caso.

In [ ]:
treino['FireplaceQu']=treino['FireplaceQu'].fillna(treino['FireplaceQu'].mode()[0])

## NaNs da coluna Electrical:
A coluna Electrical também é composta por strings.

In [ ]:
print("Quantos NaNs? {}\n".format(treino['Electrical'].isnull().sum()))
#Que registro é esse?
print(treino[treino['Electrical'].isna()])

Quantos NaNs? 1

      MSSubClass MSZoning  LotFrontage  ...  SaleType SaleCondition SalePrice
Id                                      ...                                  
1380          80       RL         73.0  ...        WD        Normal    167500

[1 rows x 76 columns]


Vemos que o registro n⁰1380 é o único NaN da Eletrical, então vamos removê-lo.

In [ ]:
treino = treino.drop(1380)

### O resto dos NaNs

Todos os outros NaNs vão ser preenchidos com a moda, pois não tem como fazer valor médio de strings.

In [ ]:
#saber quais colunas tem NaNs:
print(treino.columns[treino.isnull().any()])

Index(['MasVnrType', 'MasVnrArea', 'BsmtQual', 'BsmtCond', 'BsmtExposure',
       'BsmtFinType1', 'BsmtFinType2', 'GarageType', 'GarageYrBlt',
       'GarageFinish', 'GarageQual', 'GarageCond'],
      dtype='object')


In [ ]:
treino['MasVnrArea']=treino['MasVnrArea'].fillna(treino['MasVnrArea'].mode()[0])
treino['BsmtQual']=treino['BsmtQual'].fillna(treino['BsmtQual'].mode()[0])
treino['GarageYrBlt']=treino['GarageYrBlt'].fillna(treino['GarageYrBlt'].mode()[0])
treino['MasVnrType']=treino['MasVnrType'].fillna(treino['MasVnrType'].mode()[0])
treino['Electrical']=treino['Electrical'].fillna(treino['Electrical'].mode()[0])
treino['BsmtCond']=treino['BsmtCond'].fillna(treino['BsmtCond'].mode()[0])
treino['BsmtExposure']=treino['BsmtExposure'].fillna(treino['BsmtExposure'].mode()[0])
treino['BsmtFinType1']=treino['BsmtFinType1'].fillna(treino['BsmtFinType1'].mode()[0])
treino['GarageType']=treino['GarageType'].fillna(treino['GarageType'].mode()[0])
treino['GarageYrBlt']=treino['GarageYrBlt'].fillna(treino['GarageYrBlt'].mode()[0])
treino['GarageType']=treino['GarageType'].fillna(treino['GarageType'].mode()[0])
treino['GarageFinish']=treino['GarageFinish'].fillna(treino['GarageFinish'].mode()[0])
treino['GarageQual']=treino['GarageQual'].fillna(treino['GarageQual'].mode()[0])
treino['GarageCond']=treino['GarageCond'].fillna(treino['GarageCond'].mode()[0])
treino['BsmtFinType2']=treino['BsmtFinType2'].fillna(treino['BsmtFinType2'].mode()[0])
print(treino.columns[treino.isnull().any()])

Index([], dtype='object')


## **Fazendo a mesma coisa para os dados de teste:**

Agora que os dados de treino estão relativamente limpos e usáveis, vemos que os dados de teste também tem NaN.

In [ ]:
pd.set_option('display.max_rows', None) # mostrar todas as linhas
teste.isnull().sum()

MSSubClass          0
MSZoning            4
LotFrontage       227
LotArea             0
Street              0
Alley            1352
LotShape            0
LandContour         0
Utilities           2
LotConfig           0
LandSlope           0
Neighborhood        0
Condition1          0
Condition2          0
BldgType            0
HouseStyle          0
OverallQual         0
OverallCond         0
YearBuilt           0
YearRemodAdd        0
RoofStyle           0
RoofMatl            0
Exterior1st         1
Exterior2nd         1
MasVnrType         16
MasVnrArea         15
ExterQual           0
ExterCond           0
Foundation          0
BsmtQual           44
BsmtCond           45
BsmtExposure       44
BsmtFinType1       42
BsmtFinSF1          1
BsmtFinType2       42
BsmtFinSF2          1
BsmtUnfSF           1
TotalBsmtSF         1
Heating             0
HeatingQC           0
CentralAir          0
Electrical          0
1stFlrSF            0
2ndFlrSF            0
LowQualFinSF        0
GrLivArea 

In [ ]:
pd.reset_option('display.max_rows') # parar de mostrar todas as linhas

teste = teste.drop('Alley',axis=1)
teste = teste.drop('PoolQC',axis=1)
teste = teste.drop('Fence',axis=1)
teste = teste.drop('MiscFeature',axis=1)

teste['MasVnrType']=teste['MasVnrType'].fillna(teste['MasVnrType'].mode()[0])
teste['MasVnrArea']=teste['MasVnrArea'].fillna(teste['MasVnrArea'].mode()[0])
teste['BsmtQual']=teste['BsmtQual'].fillna(teste['BsmtQual'].mode()[0])
teste['BsmtCond']=teste['BsmtCond'].fillna(teste['BsmtCond'].mode()[0])
teste['BsmtExposure']=teste['BsmtExposure'].fillna(teste['BsmtExposure'].mode()[0])
teste['BsmtFinType1']=teste['BsmtFinType1'].fillna(teste['BsmtFinType1'].mode()[0])
teste['GarageType']=teste['GarageType'].fillna(teste['GarageType'].mode()[0])
teste['GarageYrBlt']=teste['GarageYrBlt'].fillna(teste['GarageYrBlt'].mode()[0])
teste['GarageType']=teste['GarageType'].fillna(teste['GarageType'].mode()[0])
teste['GarageYrBlt']=teste['GarageYrBlt'].fillna(teste['GarageYrBlt'].mode()[0])
teste['GarageFinish']=teste['GarageFinish'].fillna(teste['GarageFinish'].mode()[0])
teste['GarageQual']=teste['GarageQual'].fillna(teste['GarageQual'].mode()[0])
teste['GarageCond']=teste['GarageCond'].fillna(teste['GarageCond'].mode()[0])
teste['BsmtFinType2']=teste['BsmtFinType2'].fillna(teste['BsmtFinType2'].mode()[0])
teste['MSZoning']=teste['MSZoning'].fillna(teste['MSZoning'].mode()[0])
teste['Utilities']=teste['Utilities'].fillna(teste['Utilities'].mode()[0])
teste['Exterior1st']=teste['Exterior1st'].fillna(teste['Exterior1st'].mode()[0])
teste['Exterior2nd']=teste['Exterior2nd'].fillna(teste['Exterior2nd'].mode()[0])
teste['GarageCars']=teste['GarageCars'].fillna(teste['GarageCars'].mode()[0])
teste['BsmtHalfBath']=teste['BsmtHalfBath'].fillna(teste['BsmtHalfBath'].mode()[0])
teste['KitchenQual']=teste['KitchenQual'].fillna(teste['KitchenQual'].mode()[0])
teste[ 'Functional']=teste[ 'Functional'].fillna(teste[ 'Functional'].mode()[0])
teste['FireplaceQu']=teste['FireplaceQu'].fillna(teste['FireplaceQu'].mode()[0])
teste['GarageArea']=teste['GarageArea'].fillna(teste['GarageArea'].mode()[0])
teste['SaleType']=teste['SaleType'].fillna(teste['SaleType'].mode()[0])
teste['BsmtFinSF1']=teste['BsmtFinSF1'].fillna(teste['BsmtFinSF1'].mode()[0])
teste['BsmtFinSF2']=teste['BsmtFinSF2'].fillna(teste['BsmtFinSF2'].mode()[0])
teste['BsmtUnfSF']=teste['BsmtUnfSF'].fillna(teste['BsmtUnfSF'].mode()[0])
teste['TotalBsmtSF']=teste['TotalBsmtSF'].fillna(teste['TotalBsmtSF'].mode()[0])
teste['BsmtFullBath']=teste['BsmtFullBath'].fillna(teste['BsmtFullBath'].mode()[0])
print(teste.columns[teste.isnull().any()])

Index(['LotFrontage'], dtype='object')


In [ ]:
'''No dataframe de teste 'LotFrontage' também tem NaNs suficientes para ser a maior coluna numérica de dados 
faltantes, mas não tem NaNs suficientes para considerarmos tirar a coluna. Usamos KNN novamente.'''

#vamos usar o mesmo imputador já criado
teste["LotFrontage"] = imp.fit_transform(teste[["LotFrontage"]]).ravel()

print("Temos NaNs? {}\n".format(treino['LotFrontage'].isnull().any()))

Temos NaNs? False



## **Tipagem dos dados:**

Mesmo com todos os dados dos dois dataframes presentes sendo válidos, as suas colunas ainda estão misturadas em dados numéricos e categóricos; contínuos e discretos; floats, inteiros e strings.

Como todas as colunas dos dois datasets necessitarão ser separadas por tipo, é eficiente juntarmos os dois para manipulação única. 

In [ ]:
#Vemos o tamanho de cada df:
print(treino.shape)
print(teste.shape) # o teste só tem 75 colunas porque não tem o preço da casa, que é o que procuramos.

(1459, 76)
(1459, 75)


#### **Juntando datasets**

É mais lógico juntar os dados para continuar o processamento do que fazer as mesmas coisas duas vezes. 
Para isso criamos um dataframe com todos os dados usando a função .concat():

In [ ]:
df_total = pd.concat([treino,teste],ignore_index=True)
print(df_total.shape) #sabendo que: os primeiros 1459 registros vem do dados, e os 1459 últimos vem do teste.

(2918, 76)


## **Colunas numéricas e colunas categóricas:**
Iremos dividir o nosso dataframe total em colunas numéricas e categóricas para trabalhar com cada uma separadamente.

Para isso temos a função [ .select_dtypes](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.select_dtypes.html), que como já diz, seleciona as colunas com o tipo do argumento.

In [ ]:
numericos = df_total.select_dtypes(np.number)
col_numer = numericos.columns
print(col_numer)
print("\nSão {} colunas numéricas.".format(len(col_numer)))

Index(['MSSubClass', 'LotFrontage', 'LotArea', 'OverallQual', 'OverallCond',
       'YearBuilt', 'YearRemodAdd', 'MasVnrArea', 'BsmtFinSF1', 'BsmtFinSF2',
       'BsmtUnfSF', 'TotalBsmtSF', '1stFlrSF', '2ndFlrSF', 'LowQualFinSF',
       'GrLivArea', 'BsmtFullBath', 'BsmtHalfBath', 'FullBath', 'HalfBath',
       'BedroomAbvGr', 'KitchenAbvGr', 'TotRmsAbvGrd', 'Fireplaces',
       'GarageYrBlt', 'GarageCars', 'GarageArea', 'WoodDeckSF', 'OpenPorchSF',
       'EnclosedPorch', '3SsnPorch', 'ScreenPorch', 'PoolArea', 'MiscVal',
       'MoSold', 'YrSold', 'SalePrice'],
      dtype='object')

São 37 colunas numéricas.


In [ ]:
categoricos = df_total.select_dtypes(exclude= np.number)
col_categ = categoricos.columns
print(col_categ)
print("\nSão {} colunas categóricas.".format(len(col_categ)))

Index(['MSZoning', 'Street', 'LotShape', 'LandContour', 'Utilities',
       'LotConfig', 'LandSlope', 'Neighborhood', 'Condition1', 'Condition2',
       'BldgType', 'HouseStyle', 'RoofStyle', 'RoofMatl', 'Exterior1st',
       'Exterior2nd', 'MasVnrType', 'ExterQual', 'ExterCond', 'Foundation',
       'BsmtQual', 'BsmtCond', 'BsmtExposure', 'BsmtFinType1', 'BsmtFinType2',
       'Heating', 'HeatingQC', 'CentralAir', 'Electrical', 'KitchenQual',
       'Functional', 'FireplaceQu', 'GarageType', 'GarageFinish', 'GarageQual',
       'GarageCond', 'PavedDrive', 'SaleType', 'SaleCondition'],
      dtype='object')

São 39 colunas categóricas.


## **Processamento dos dados categóricos**

Um dos métodos para vetorizar os dados dessas colunas é o [LabelBinarizer](https://scikit-learn.org/stable/modules/generated/sklearn.preprocessing.LabelBinarizer.html), mas com apenas uma coluna pelo método one-vs-all. 

Para fazer múltiplas colunas de uma vez podemos utilizar o [.get_dummies](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.get_dummies.html) e mudar os parâmetros para colunas, uma vez que o padrão são linhas.

In [ ]:
df_preparado = pd.get_dummies(df_total, columns=col_categ)
df_preparado.shape
#são 37 colunas originalmente numéricas, e 39 colunas transformaram-se em 239 colunas geradas pelo .get_dummies.

#Entretetanto,agora temos 2 de cada coluna categórica, a normal e a vetorizada.
# Podemos tirar isso com a drop_duplicates()
print(df_preparado.shape)
df_preparado.drop_duplicates()
df_preparado.head()

(2918, 276)


MSSubClass  LotFrontage  ...  SaleCondition_Normal  SaleCondition_Partial
0          60         65.0  ...                     1                      0
1          20         80.0  ...                     1                      0
2          60         68.0  ...                     1                      0
3          70         60.0  ...                     0                      0
4          60         84.0  ...                     1                      0

[5 rows x 276 columns]

Agora temos um dataframe único com todas as colunas numéricas. Exceto, é claro, a coluna 'SalePrice'. 

Os primeiros 1459 registros possuem algum valor(vindos do dataset dados), e os outros 1459 possuem NaN(vindos do dataset teste).

In [ ]:
#Dados rotulados da 'SalePrice' vindos do df treino
print(df_preparado['SalePrice'].head())
#Dados rotulados da 'SalePrice' vindos do df teste
print(df_preparado['SalePrice'].tail())

0    208500.0
1    181500.0
2    223500.0
3    140000.0
4    250000.0
Name: SalePrice, dtype: float64
2913   NaN
2914   NaN
2915   NaN
2916   NaN
2917   NaN
Name: SalePrice, dtype: float64


Como o dataset que vai ser usado para o treinamento é a partição que contém a coluna 'SalePrice', e como nós só juntamos os datasets para questão de uniformidade de OneHot, podemos redividí-los em treino e teste.

In [ ]:
#Vamos fazer uma cópia das primeiras 1459 linhas para o df_treino
df_treino=df_preparado.iloc[:1459,:].copy()
#Vamos fazer uma cópia das últimas 1459 linhas para o df_teste
df_teste=df_preparado.iloc[1459:,:].copy()

#Como foi criado uma coluna 'SalePrice' no dataframe de teste quando juntamos os datasets, vamos retirá-la:
df_teste.drop(['SalePrice'],axis=1,inplace=True)

## **Treinamento do modelo**

In [ ]:
#vamos criar os dois parâmetros que a função fit recebe:
xtreino = df_treino.drop(['SalePrice'],axis=1)
# O segundo parâmetro que indica os rótulos:
ytreino = df_treino['SalePrice']

Separa os dados de treino e teste.

In [ ]:
from sklearn.model_selection import train_test_split
x_train, x_test, y_train, y_test = train_test_split(df_treino, df_teste, test_size = 0.25, random_state=42)

Implementa o modelo Multi-Layer Perceptron (MLP), treinando com o algoritmo Backpropagation.

In [ ]:
from sklearn.neural_network import MLPRegressor
mlp = MLPRegressor(random_state=1, max_iter=1000)
modelo_treinado = mlp.fit(xtreino, ytreino)

print(modelo_treinado)

expected_y  = y_test
predicted_y = modelo_treinado.predict(xtreino)

MLPRegressor(activation='relu', alpha=0.0001, batch_size='auto', beta_1=0.9,
             beta_2=0.999, early_stopping=False, epsilon=1e-08,
             hidden_layer_sizes=(100,), learning_rate='constant',
             learning_rate_init=0.001, max_fun=15000, max_iter=1000,
             momentum=0.9, n_iter_no_change=10, nesterovs_momentum=True,
             power_t=0.5, random_state=1, shuffle=True, solver='adam',
             tol=0.0001, validation_fraction=0.1, verbose=False,
             warm_start=False)


## **Avaliação do modelo**

Para avaliar a quantidade de erro do nosso modelo podemos utilizar o RMSE (root mean squared error) no dados de treino. Uma vez que não podemos calcular a acurácia nos dados de teste porque não temos os rótulos.

In [ ]:
from sklearn.metrics import mean_squared_error
p = modelo_treinado.predict(xtreino)

print("Quantidade de erros: {}".format(np.sqrt(mean_squared_error(ytreino, p))))

Quantidade de erros: 40634.99519602094


Com isso, podemos voltar no nosso regressor e ajustar alguns de seus parâmetros manualmente, de forma a diminuir esse erro.

In [ ]:
##criamos nosso regressor:
regressor2= MLPRegressor(
                           hidden_layer_sizes=(20,20,20), #três camadas ocultas são modeladas, com 20 neurônios em cada camada
                           activation="relu", #função de ativação para a camada oculta
                           random_state=1, 
                           max_iter=5000).fit(xtreino, ytreino
                            )#aplicamos o regressor no modelo.

modelo_treinado2 = regressor2

p = modelo_treinado2.predict(xtreino)

print("Quantidade de erros: {}".format(np.sqrt(mean_squared_error(ytreino, p))))

Quantidade de erros: 22653.017460805575


In [ ]:
#Vamos comparar as predições:
predicao1 = modelo_treinado.predict(df_teste)
print("Predição1:",predicao1)
predicao2 = modelo_treinado2.predict(df_teste)
print("Predição2:",predicao2)

Predição1: [153121.89608838 174181.62633864 197442.43878662 ... 203295.35670122
  94155.14383126 238333.41640677]
Predição2: [127128.11438573 167979.98526571 190089.70085904 ... 173287.68783029
 121041.64995373 212133.07243664]


In [ ]:
pred = pd.DataFrame({'Id':teste.index, 'SalePrice': predicao2})

print(pred)
pred.to_csv('predicao2.csv', index=False)

        Id      SalePrice
0     1461  127128.114386
1     1462  167979.985266
2     1463  190089.700859
3     1464  193611.866046
4     1465  183578.199648
...    ...            ...
1454  2915   88450.802678
1455  2916   90328.887587
1456  2917  173287.687830
1457  2918  121041.649954
1458  2919  212133.072437

[1459 rows x 2 columns]
